### ISSUE TITLE
RetryAnalyzer to rerun failed tests (with specific input that failed) with dataproviders in parallel

### ISSUE THREAD

**Issue by baekkyok** [cbeust/testng#1166]
### TestNG Version

TestNG-6.x.103.1429
### Expected behavior

I currently have a dataprovider that has multiple input objects run in parallel by the test class

`@DataProvider(parallel = true)`

My retry logic looks like this:

``` java
import org.testng.IRetryAnalyzer;
import org.testng.ITestResult;

public class RetryAnalyzer implements IRetryAnalyzer  {
    private int count = 0;
    private int maxCount = 2;

    @Override
    public boolean retry(ITestResult result) {
        if(count < maxCount && !result.isSuccess()) {
            log.debug("Going to retry test case: " + result.getMethod() + ", " + count + " out of " + maxCount);
            count++;
            return true;
        }
        return false;
    }
}
```

I have a single test class that has:

`@Test(timeOut = MAX_TIMEOUT, retryAnalyzer = RetryAnalyzer.class)`

My expectation was that if the test failed for 'input TYPE1', it would run the test class with 'input TYPE1' up to  maxCount = 2.
### Actual behavior

However, based on test output logs, what is happening is the test is rerun 3 times for three random inputs that failed. 

Ex.

Test failed at  
`'input TYPE2', 'input TYPE4', 'input TYPE5', 'input TYPE8', 'input TYPE9'`

rather than retrying each of them 3 times for the same class, it is retrying once for only 3 input cases.
`'input TYPE2', 'input TYPE4', 'input TYPE5'.`

 The result shows those 3 retrys as 3 skipped tests at the end.

Is this a correct behavior? 
Please let me know if I am missing something.

Thank you for your help.
### Is the issue reproductible on runner?
- [x] Shell

**Comment by juherr** [comment]
> TestNG-6.x.103.1429

I don't know what is this version. Where do you find it?

About your issue, I suppose retry analyser is considering a method and not a method and its parameters. It looks wrong but I can't say if a fix won't break things.
The random part is due to parallelization.

As a workaround, you can use a `@Factory` linked to your `@DataProvider` and remove it from your test method.
TestNG will create 1 instance by data provider case and will consider each method+instance as unique. RetryAnalyzer should work in this case.

**Comment by baekkyok** [comment]
Hi juherr,

Thank you for your reply.

I currently have @Factory on my constructor for the test class

` @Factory(dataProvider = "getHttpInputs", dataProviderClass = HttpDataProvider.class)
        public IntegrationTest(List<sInput> sInputs, List<pInput> pInputs) {
    `

`@Test(groups = {"integration", "validation"}, timeOut = MAX_TIMEOUT, retryAnalyzer = RetryAnalyzer.class)
        public void test() throws Exception {`

Are you suggesting moving RetryAnalyzer to the constructor?

` @Factory(dataProvider = "getHttpInputs", dataProviderClass = HttpDataProvider.class,  retryAnalyzer = RetryAnalyzer.class)
        public IntegrationTest(List<sInput> sInputs, List<pInput> pInputs) {
    `

`@Test(groups = {"integration", "validation"}, timeOut = MAX_TIMEOUT)
        public void test() throws Exception {`


Additionally, if we want to create 1 instance by data provider case and will consider each method+instance as unique. Should I change parallel="methods" to  parallel="instances"
?

`<suite name="IntegrationTests" verbose="2" parallel="methods" thread-count="20" data-provider-thread-count="20">
`

**Comment by baekkyok** [comment]
I have tried  parallel="instances" in testng.xml, did not help.
I have tried adding `retryAnalyzer = RetryAnalyzer.class` to `@Factory` like this..
`@Factory(dataProvider = "geInputs", dataProviderClass = DataProvider.class, retryAnalyzer = RetryAnalyzer.class) public IntegrationTest(List<sInput> sInputs, List<pInput> pInputs) { `

but `@Factory` did not recognize retryAnalyzer class. 

Could you explain in more detail what you mean by 

> It looks wrong but I can't say if a fix won't break things.

> As a workaround, you can use a @Factory linked to your @DataProvider and remove it from your test method.


TestNG creating 1 instance by data provider case and will consider each method+instance as unique is exactly what I want in my case to retry failed tests with specific inputs.

Thank you.

**Comment by juherr** [comment]
Could you provide a full sample? It's a bit hard to understand without.

**Comment by baekkyok** [comment]
Sorry, I am not able to provide a full sample.

I will try to give you as much context as possible.

**RetryAnalyzer.java**
`import org.testng.IRetryAnalyzer;
import org.testng.ITestResult;

public class RetryAnalyzer implements IRetryAnalyzer  {
    private int count = 0;
    private int maxCount = 2;

    @Override
    public boolean retry(ITestResult result) {
        if(count < maxCount && !result.isSuccess()) {
            log.debug("Going to retry test case: " + result.getMethod() + ", " + count + " out of " + maxCount);
            count++;
            return true;
        }
        return false;
    }
}`

**IntegrationTest.java**
`public class IntegrationTest extends TestBase implements ITest {

   ....
    private static final Logger log = new Logger();

    private final List<SubscribeInput> subscribeInputs;
    private final List<PublishInput> publishInputs;

    @Factory(dataProvider = "getInputs", dataProviderClass = DataProvider.class)
        public IntegrationTest(List<SubscribeInput> subscribeInputs, List<PublishInput> publishInputs) {
            Assert.assertTrue(subscribeInputs != null);
            Assert.assertTrue(publishInputs != null);
            this.subscribeInputs = subscribeInputs;
            this.publishInputs = publishInputs;
        }

    @Override
        public String getTestName() {
            if (subscribeInputs.size() > 0) {
                return subscribeInputs.get(0).getTopic();
            } else if (publishInputs.size() > 0) {
                return publishInputs.get(0).getTopic();
            } else {
                return "test";
            }
        }

    @Test(groups = {"integration", "validation"}, timeOut = MAX_TIMEOUT, retryAnalyzer = RetryAnalyzer.class)
        public void test() throws Exception {
        ......... test content....
        .................................

`

**DataProvider.java**
`public class DataProvider extends AbstractDataProvider {

    private static final Logger log = new Logger();

    ...................

    @DataProvider(name = "getInputs", parallel = true)
    public static Object[][] getInputs() throws Exception {
        List<Object[]> allInputs = getInputA();
        allInputs.addAll(getInputB());
        allInputs.addAll(getInputC());
        allInputs.addAll(getInputD());
        return allInputs.toArray(new Object[allInputs.size()][]);
    }`


**testng.xml**
`<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >

<suite name="MyIntegrationTests" verbose="2" parallel="instances" thread-count="20" data-provider-thread-count="20">
	<test name="IntegrationTests">
		<groups>
			<run>
				
				
			</run>
		</groups>
		<packages>
			<package name="integration.*" />
		</packages>
	</test>
</suite>`

**Comment by juherr** [comment]
It is full enough, thanks :)

**Comment by baekkyok** [comment]
Hi, is there any recommendations to achieve desired behavior of TestNG? Thanks.

**Comment by juherr** [comment]
Not for the moment unless you try to fix it by yourself.

**Comment by baekkyok** [comment]
I see. Do you mean the observed behavior of the tesNG (running each failed tests only once for number of tests up to selected # of retries in the RetryAnalyzer class) is normal and expected?

Could you explain in more detail about your previous comments?

'It looks wrong but I can't say if a fix won't break things.'
What is the fix?

and

'As a workaround, you can use a @Factory linked to your @DataProvider and remove it from your test method.'
What is "it" you are referring to here?

I appreciate your help

**Comment by juherr** [comment]
> normal &  expected

I can't say.  @cbeust ? 

> Fix ? 

No real idea for the moment. Maybe something around test method registration. 

> it

I supposed you have a data provider on the test method but it is not the case.

**Comment by SrinivasanTarget** [comment]
@juherr Any update on this? Will give a try to fix if you could point me in code.

**Comment by krmahadevan** [comment]
@SrinivasanTarget - The easiest way of getting started with fixing this issue would be leverage the sample that has been shared here, setup break-points in the codebase and start walking through the code to see where the breakage is. Atleast that is the normal approach that I have been doing when it comes to fixing issues in TestNG. Have you tried that ? If yes, where are you stuck. Maybe you could call it out. I am not sure if @juherr would have the bandwidth to provide pointers to the relevant sections that need to be looked at for fixing this issue (or any testng issue in general) [ considering the amount of things he is involved with ]

**Comment by juherr** [comment]
As @krmahadevan said, I've not enough free time for the moment. But I can help to review pull requests and to merge them.

**Comment by jcdkount** [comment]
I have been trying to understand why we would occasionally see a test execute multiple times.  Eventually I happened upon this bug.  In tracking it down, I captured a few pieces of information that are not in the current bug.  All of my conclusions can be checked by my test project: [bug-1166.zip](https://github.com/cbeust/testng/files/784203/bug-1166.zip)


1. Once a data provider test fails on a particular value, all other values for that test will "retry" 1+ times, depending on how many different values ultimately fail.
2. Once a data provider test fails on a particular value, and other values retry, the stack trace suggests the code thinks it is retrying a failure.
3. The retry analyzer is NOT invoked on the second execution of a previously passing value.  That is to say, once a data provider test fails on a particular value, and other values "retry", they do not query the retry analyzer.

I also found that the code path seems to vary once a failure occurs.  I used the TestListenerAdapter’s events to capture the stack and I think there is some interesting data points.  I should note these stack traces are from 6.9.9 but they also break in 6.9.10.

First value that is a failure (turns to a skip due to being retried):

at test.TestNGTestListenerAdapter.onTestSkipped(TestNGTestListenerAdapter.java:18)
at org.testng.internal.Invoker.runTestListeners(Invoker.java:1678)
at org.testng.internal.Invoker.runTestListeners(Invoker.java:1670)
at org.testng.internal.Invoker.invokeMethod(Invoker.java:696)
at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:816)
at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1124)
at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
at java.lang.Thread.run(Thread.java:745)

Then when it retries, it is successful:

at test.TestNGTestListenerAdapter.onTestSuccess(TestNGTestListenerAdapter.java:13)
at org.testng.internal.Invoker.runTestListeners(Invoker.java:1690)
at org.testng.internal.Invoker.runTestListeners(Invoker.java:1670)
at org.testng.internal.Invoker.invokeMethod(Invoker.java:696)
**at org.testng.internal.Invoker.retryFailed(Invoker.java:963)**
at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1142)
at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
at java.lang.Thread.run(Thread.java:745)

The first pass of any value used after a failing value via the data provider test:

at test.TestNGTestListenerAdapter.onTestSuccess(TestNGTestListenerAdapter.java:13)
at org.testng.internal.Invoker.runTestListeners(Invoker.java:1690)
at org.testng.internal.Invoker.runTestListeners(Invoker.java:1670)
at org.testng.internal.Invoker.invokeMethod(Invoker.java:696)
at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:816)
at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1124)
at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
at java.lang.Thread.run(Thread.java:745)

A "retry" for a test that passed the first time.  Note that it hit "retryFailed" as if the previous test had failed:

at test.TestNGTestListenerAdapter.onTestSuccess(TestNGTestListenerAdapter.java:13)
at org.testng.internal.Invoker.runTestListeners(Invoker.java:1690)
at org.testng.internal.Invoker.runTestListeners(Invoker.java:1670)
at org.testng.internal.Invoker.invokeMethod(Invoker.java:696)
**at org.testng.internal.Invoker.retryFailed(Invoker.java:963)**
at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1142)
at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:125)
at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
at java.lang.Thread.run(Thread.java:745)

While I didn't dig too deeply into the source code, I would speculate that [failure.instance](https://github.com/cbeust/testng/blob/6.9.9/src/main/java/org/testng/internal/Invoker.java#L1136) is not parameter aware and so it thinks the method has failed and should be retried.  Since the same method is used for each parameter value, it retries them all.  It probably needs to be parameter aware.

**Comment by vipin0060** [comment]
Hi, 

I am facing the same issue now as well in 6.9.10. Can anyone please suggest me a workaround for this issue ?

**Comment by alexanderm2018** [comment]
same issue for me
![image](https://user-images.githubusercontent.com/36869564/40943634-b9447020-6806-11e8-8efb-053805c099d7.png)

**Comment by alexanderm2018** [comment]
Facing the same issue with testng version 6.14.12
![image](https://user-images.githubusercontent.com/36869564/40943709-fee55d10-6806-11e8-84d2-a425995a6177.png)

**Comment by krmahadevan** [comment]
There have been a couple of things that have been done to fix the discrepancy with retry analyzer in the `master` branch which should now be available as part of TestNG `7.0.0` (already available as part of `7.0.0-SNAPSHOT`)

* Ability to distinguish between skipped tests (status as `SKIP`) vs retried tests(status as `SKIP` with retried flag as `true` which is accessible via `ITestResult.wasRetried()`)
* TestNG provided default reports explicitly calling out `retried` tests in them.
* Now the `testing-results.xml` looks like below

```xml
<testng-results skipped="0" failed="1" ignored="0" retried="2" total="4" passed="1">
```

Please retry using TestNG `7.0.0` and comment if there's any additional problems.
